packages/lesswrong/server/scripts/makeMigrations.ts

```diff
@@ -55,6 +56,234 @@ const schemaFileHeaderTemplate = `-- GENERATED FILE
 --
 `
 
+declare global {
+  type SchemaDependency =
+    {type: "extension", name: PostgresExtension} |
+    {type: "collection", name: CollectionNameString} |
+    {type: "function", name: string} |
+    {type: "view", name: string};
+}
+
+abstract class Node {
+  public dependencies: SchemaDependency[] = [];
+
+  addDependency(dependency: SchemaDependency) {
+    this.dependencies.push(dependency);
+  }
+
+  abstract getName(): string;
+  abstract getQuery(): {compile(): {sql: string, args: any[]}};
+
+  getHeader(): string {
+    const type = this.constructor.name.replace(/(_|Node)/g, "");
+    return `-- ${type} "${this.getName()}", hash ${this.getHash()}`;
+  }
+
+  getSource(): string {
+    const {sql, args} = this.getQuery().compile();
+    return sqlInterpolateArgs(sql, args).trim();
+  }
+
+  getAnnotatedSource(): string {
+    const source = this.getSource();
+    const hasSemi = source[source.length - 1] === ";";
+    return `${this.getHeader()}\n${source}${hasSemi ? "" : ";"}\n`;
+  }
+
+  getHash(): string {
+    return md5(this.getSource());
+  }
+}
+
+class ExtensionNode extends Node {
+  constructor(private extension: PostgresExtension) {
+    super();
+  }
+
+  getName() {
+    return this.extension;
+  }
+
+  getQuery() {
+    return new CreateExtensionQuery(this.extension);
+  }
+}
+
+class TableNode extends Node {
+  constructor(private table: Table<DbObject>) {
+    super();
+  }
+
+  getName() {
+    return this.table.getName();
+  }
+
+  getQuery() {
+    return new CreateTableQuery(this.table);
+  }
+}
+
+class IndexNode extends Node {
+  constructor(
+    private table: Table<DbObject>,
+    private index: TableIndex<DbObject>,
+  ) {
+    super();
+    this.addDependency({
+      type: "collection",
+      name: table.getName() as CollectionNameString,
+    });
+  }
+
+  getName() {
+    return this.index.getName();
+  }
+
+  getQuery() {
+    return new CreateIndexQuery(this.table, this.index);
+  }
+}
+
+class CustomIndexNode extends Node {
+  private static nameRegex = /^\s*CREATE\s+(UNIQUE\s+)?INDEX\s+(CONCURRENTLY\s+)?(IF\s+NOT\s+EXISTS\s+)?"?([a-zA-Z0-9_]+)/i;
```

Comment:
This is far from fool-proof, but should be easy enough to fix if it breaks in the future